### PR TITLE
Fix penalties and add tests

### PR DESCRIPTION
**Pull request by scarlehoff**
These have been broken since a while, but since nobody used them it didn't really matter.

Don't think they will be used but who knows, I added a few tests since (even if they are not used in practice) they ensure that several pieces of the code work well together as they require the correct flavours being set, `n3fit` model to be compatible with validphys, etc.

### CODE REVIEW

File: n3fit/src/n3fit/tests/test_penalties.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+"""
+    Test the penalties for n3fit hyperopt
+"""
+from types import SimpleNamespace
+from n3fit.hyper_optimization.penalties import integrability, patience, saturation
+from n3fit.model_gen import pdfNN_layer_generator
+
+
+def test_saturation():
+    """Check the saturation penalty runs and returns a float"""
```
Comment (RoyStegeman) [inline n3fit/src/n3fit/tests/test_penalties.py:10]:
```suggestion
    """Check that the saturation penalty runs and returns a float"""
```

File: n3fit/src/n3fit/tests/test_penalties.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    fake_stopping = SimpleNamespace(
+        e_best_chi2=1000, stopping_patience=500, total_epochs=5000, vl_loss=2.42
+    )
+    res = patience(stopping_object=fake_stopping, alpha=1e-4)
+    assert isinstance(res, float)
+
+
+def test_integrability_numbers():
+    """Check that the integrability penality can be run
+    and that it returns indeed a float"""
```
Comment (RoyStegeman) [inline n3fit/src/n3fit/tests/test_penalties.py:32]:
```suggestion
    """Check that the integrability penalty runs and returns a float"""
```

**Review by RoyStegeman — COMMENTED** [review]
I don't think anyone is using this, though @Zaharid did contact me about it recently so he may want to be aware of this.

File: n3fit/src/n3fit/hyper_optimization/penalties.py [hunk 3]
```diff
@@ -82,15 +82,15 @@ def patience(stopping_object=None, alpha=1e-4, **_kwargs):
...
+    >>> patience(fake_stopping, alpha=1e-4)
     3.434143467595683
 
     """
     epoch_best = np.take(stopping_object.e_best_chi2, 0)
     patience = stopping_object.stopping_patience
     max_epochs = stopping_object.total_epochs
     diff = abs(max_epochs - patience - epoch_best)
-    vl_loss = np.take(stopping_object.vl_chi2, 0)
+    vl_loss = np.take(stopping_object.vl_loss, 0)
```
Comment (RoyStegeman) [inline n3fit/src/n3fit/hyper_optimization/penalties.py:93]: I don't believe this is correct

Comment (scarlehoff) [reply n3fit/src/n3fit/hyper_optimization/penalties.py:93]: What do you mean? In principle given the description it should be `loss` (so before it was wrong).

Comment (RoyStegeman) [reply n3fit/src/n3fit/hyper_optimization/penalties.py:93]: Because it's a `Stopping` object, not a `Fitstate` object (though I guess it might actually make life easier to just make it one object). The former does not have the `vl_loss` property.

Comment (scarlehoff) [inline n3fit/src/n3fit/hyper_optimization/penalties.py:93]:
```suggestion
    vl_loss = np.take(stopping_object.vl_chi2, 0)
```

I see. Then let me revert that.

File: n3fit/src/n3fit/tests/test_penalties.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    pdf_model = pdfNN_layer_generator(
+        nodes=[8], activations=["linear"], seed=0, flav_info=fake_fl, fitbasis="FLAVOUR"
+    )
+    assert isinstance(saturation(pdf_model, 5), float)
+
+
+def test_patience():
+    """Check that the patience penalty runs and returns a float"""
+    fake_stopping = SimpleNamespace(
+        e_best_chi2=1000, stopping_patience=500, total_epochs=5000, vl_loss=2.42
```
Comment (scarlehoff) [inline n3fit/src/n3fit/tests/test_penalties.py:24]:
```suggestion
        e_best_chi2=1000, stopping_patience=500, total_epochs=5000, vl_chi2=2.42
```

**Review by RoyStegeman — APPROVED** [review]

**Comment by RoyStegeman** [comment]
I think this is ready to be merged?
